lib/ecto/association.ex

```diff
@@ -176,97 +175,163 @@ defmodule Ecto.Association do
   end
 
   @doc """
-  Build an association query through with starting the given reflection
-  and through the given associations.
+  Build an association query through the given associations from the specified owner table
+  and through the given associations. Finally filter by the provided values of the owner_key of
+  the first relationship in the chain. Used in Ecto.assoc/2.
   """
-  def assoc_query(refl, through, query, values)
-
-  def assoc_query(%{owner: owner, through: [h|t], field: field}, extra, query, values) do
-    refl = owner.__schema__(:association, h) ||
-            raise "unknown association `#{h}` for `#{inspect owner}` (used by through association `#{field}`)"
-    assoc_query refl, t ++ extra, query, values
+  def filter_through_chain(owner, through, values) do
+    chain_through(owner, through, nil, values)
+    |> distinct([x], true)
```

Comment:
I'm actually confused as to why this `distinct` clause needs to be here. It's used at the end of `assoc_query/4` which is being replaced by this function. Without it certain tests fail due to duplicate records. Yet `ManyToMany` does not add on a distinct clause, and so there are integration tests which use `Ecto.assoc/2` to query the data but expect duplicate records, for example [this test case](https://github.com/elixir-ecto/ecto/blob/master/integration_test/cases/assoc.exs#L137).

I'm unclear on when and when not we should be using distinct in association queries. Perhaps this is something to make more consistent in a follow up PR?

Reply:
My understanding is that it only applies to through associations and it was there since the beginning (I added it more than 5 years ago - so I can't quite remember). :+1: to revisiting later.